src/pages/index.js

```diff
@@ -78,19 +104,129 @@ editProfileButton.addEventListener("click", () => {
   profilePopup.open();
 });
 
-//add new card
+profileImageElement.addEventListener("click", () => {
+  imageChangePopup.open();
+});
+
+/*////////////////////////////////////////////////////////////////////
+/                      Popups                                        /
+////////////////////////////////////////////////////////////////////*/
+
+const viewImagePopup = new PopupWithImage(".view-image-popup");
+const confirmDeletePopup = new PopupWithForm({
+  popupSelector: ".delete-card-popup",
+  handleFormSubmit: () => {
+    deleteCard(cardElement);
+  },
+});
 
 const addCardPopup = new PopupWithForm({
   popupSelector: ".add-card-popup",
   handleFormSubmit: (formData) => {
-    renderCard(formData);
-    addCardPopup.close();
+    addCardPopup.setSavingMessage();
+    api
+      .addCard(formData)
+      .then((data) => {
+        renderCard(data);
+        addCardPopup.close();
+        addCardPopup.resetSubmitMessage("Create");
+      })
+      .catch((err) => {
+        console.error(`Error: ${err}`);
+      });
   },
 });
 
-const addLocationButton = document.querySelector(".profile__add-button");
+const profilePopup = new PopupWithForm({
+  popupSelector: ".profile-popup",
+  handleFormSubmit: () => {
+    const data = {
+      name: editProfileName.value,
+      about: editProfileSubtitle.value,
+    };
+    profilePopup.setSavingMessage();
+    api
+      .updateUserInfo(data)
+      .then((res) => {
+        userInfo.setUserInfo({
+          name: res["name"],
+          info: res["about"],
+          image: res["avatar"],
+          id: res["_id"],
+        });
+        profilePopup.close();
+        profilePopup.resetSubmitMessage("Save");
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```

Reply:
Will address once must-haves are complete